Fixes @media and invalid test cases

leonsenft opened the pull request:
The parser now supports media queries with expressionless media features (no trailing `: <expr>`). For example:

    @media all and (transform-3d) {}

Also adds missing AND operators to test cases which had expressions immediately following the media type.

Fixes dart-lang/tools#1153.

leonsenft committed:
Fixes @media and invalid test cases

The parser now supports media queries with expressionless media features (no
trailing `: <expr>`). For example:

    @media all and (transform-3d) {}

Also adds missing AND operators to test cases which had expressions immediately
following the media type.

Fixes #44.

terrylucas commented on test/declaration_test.dart:
```diff
@@ -411,13 +411,13 @@ void testMediaQueries() {
 
```
Good catch on the repeating expressions must have an AND.  However, there are no tests for the simplest case with one expression e.g.,

@media (min-device-width: 4000px) {
...
}

```
media_query_list
 : S* [media_query [ ',' S* media_query ]* ]?
 
media_query
 : [ONLY | NOT]? S* media_type S* [ AND S* expression ]* | expression [ AND S* expression ]*

media_type
 : IDENT

expression
 : '(' S* media_feature S* [ ':' S* expr ]? ')' S*

media_feature
 : IDENT
```

Do you think its a good idea to have a failure case that is caught when there is more than one expression and no AND operator?

leonsenft replied:
Good call, I'll add tests for both.

leonsenft replied:
Done and ready for review.

terrylucas reviewed (approved):
lgtm, and good catch.  Just some comments on a few more tests.

terrylucas commented:
Reviewed with comments.

On Fri, May 5, 2017 at 6:35 PM, Leon Senft <notifications@github.com> wrote:

> @leonsenft <https://github.com/leonsenft> requested your review on:
> dart-lang/csslib#47 <https://github.com/dart-lang/csslib/pull/47> Fixes
> @media <https://github.com/media> and invalid test cases.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dart-lang/csslib/pull/47#event-1071427668>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB9ewQp5UAlLby86JyWVThkWS9XEgXNpks5r285OgaJpZM4NSk06>
> .
>

leonsenft committed:
Adds more tests and fixes query list parsing

Changing how media queries are parsed introduced a regression where parsing a
malformed media query list would consume tokens past the initial point of
failure.

terrylucas reviewed (approved):
lgtm now.
